rust/yarp/src/lib.rs

```diff
@@ -0,0 +1,322 @@
+//! # yarp
+//!
+//! Rustified version of Ruby's YARP parser.
+//!
+#![warn(
+    box_pointers,
+    clippy::all,
+    clippy::nursery,
+    clippy::pedantic,
+    future_incompatible,
+    missing_docs,
+    nonstandard_style,
+    rust_2018_idioms,
+    trivial_casts,
+    trivial_numeric_casts,
+    unreachable_pub,
+    unused_qualifications
+)]
+
+// Most of the code in this file is generated, so sometimes it generates code
+// that doesn't follow the clippy rules. We don't want to see those warnings.
+#![allow(
+    clippy::too_many_lines,
+    clippy::use_self
+)]
+
+mod bindings {
+    // In `build.rs`, we generate bindings based on the config.yml file. Here is
+    // where we pull in those bindings and make them part of our library.
+    include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+}
+
+use std::ffi::c_char;
+use std::{ffi::CStr, marker::PhantomData};
+use std::mem::MaybeUninit;
+use std::ptr::NonNull;
+
+use yarp_sys::{yp_parser_t, yp_parser_init, yp_parse, yp_node_destroy, yp_node_t, yp_comment_t, yp_diagnostic_t};
+pub use self::bindings::*;
+
+/// A diagnostic message that came back from the parser.
+#[derive(Debug)]
+pub struct Diagnostic<'pr> {
+    diagnostic: NonNull<yp_diagnostic_t>,
+    marker: PhantomData<&'pr yp_diagnostic_t>
+}
+
+impl<'pr> Diagnostic<'pr> {
+    /// Returns the message associated with the diagnostic.
+    /// 
+    /// # Panics
+    ///
+    /// Panics if the message is not able to be converted into a `CStr`.
+    ///
+    #[must_use]
+    pub fn message(&self) -> String {
+        unsafe {
+            let message: *mut c_char = self.diagnostic.as_ref().message as *mut c_char;
+            CStr::from_ptr(message).to_str().unwrap().to_string()
+        }
+    }
+}
+
+/// A comment that was found during parsing.
+#[derive(Debug)]
+pub struct Comment<'pr> {
+    comment: NonNull<yp_comment_t>,
+    marker: PhantomData<&'pr yp_comment_t>
+}
+
+impl<'pr> Comment<'pr> {
+    /// Returns the text of the comment.
+    ///
+    /// # Panics
+    /// 
+    /// Panics if the comment is not able to be read.
+    /// 
+    #[must_use]
+    pub fn text(&self) -> String {
+        unsafe {
+            let start: *mut u8 = self.comment.as_ref().start as *mut u8;
+            let end: *mut u8 = self.comment.as_ref().end as *mut u8;
+
+            let len = end as usize - start as usize;
```

Comment:
Casting a pointer to `usize` is generally to be avoided and will run afoul of Rust's pointer provinance rules. The pointer type has a rich API. Use that instead.

```suggestion
            let start = self.comment.as_ref().start.cast::<u8>();
            let end = self.comment.as_ref().end.cast::<u8>();

            let len = usize::try_from(end.offset_from(start)).expect("end should point to memory after start");
```